Microservices with oauth2 auth and ManyToOne relationship to User with services fail to compile.

<!--
- Please follow the issue template below for bug reports.
- If you have a support request rather than a bug, please use [Stack Overflow](http://stackoverflow.com/questions/tagged/jhipster) with the JHipster tag.
- For bug reports it is mandatory to run the command `jhipster info` in your project's root folder, and paste the result here.
- Tickets opened without any of these pieces of information will be **closed** without any explanation.
-->

##### **Overview of the issue**

Microservices with oauth2 auth and ManyToOne relationship to User with services fail to compile.

Given the following JDL:

```
application {
  config {
    baseName gateway,
    applicationType gateway,
    authenticationType oauth2,
    packageName com.example.gateway,
    prodDatabaseType postgresql,
    clientFramework react,
    buildTool gradle
  }
  entities A, B, C, D
}

application {
  config {
    baseName microservicea,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservicea,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities A, B
}

application {
  config {
    baseName microserviceb,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microserviceb,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities C, D
}

entity A {
    data String
}

entity B {
    data String
}

entity C {
    data String
}

entity D {
    data String
}

relationship ManyToOne {
  A{user(login)} to User
}

dto * with mapstruct
service * with serviceClass

microservice A, B with microservicea
microservice C, D with microserviceb
```

building microservicea with `./gradlew bootWar -Pprod jibDockerBuild`:

```
> Task :compileJava
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.17.Final
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/AService.java:31: error: cannot find symbol
    private final UserRepository userRepository;
                  ^
  symbol:   class UserRepository
  location: class AService
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/AService.java:33: error: cannot find symbol
    public AService(ARepository aRepository, AMapper aMapper, UserRepository userRepository) {
                                                              ^
  symbol:   class UserRepository
  location: class AService
2 errors

> Task :compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
4 actionable tasks: 4 executed
```

If `dto * with mapstruct` and `service * with serviceClass` are commented out in the jdl, the microservice will build correctly.

<!-- Explain the bug, if an error is being thrown a stack trace helps -->

##### **Motivation for or Use Case**

Discovered while trying to do filtering on an entity, which requires a service. 

Tried the okta microservices demo at: https://github.com/oktadeveloper/okta-jhipster-microservices-oauth-example

Realized that example does not use services.

The documentation at Using DTO's (https://www.jhipster.tech/using-dtos) reads that this should be valid.

It may be worth noting that changing the auth type to JWT fails compilation as well:

```
> Task :compileJava FAILED
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:32: error: cannot find symbol
    private User user;
            ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:56: error: cannot find symbol
    public User getUser() {
           ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:60: error: cannot find symbol
    public A user(User user) {
                  ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:65: error: cannot find symbol
    public void setUser(User user) {
                        ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/mapper/AMapper.java:11: error: cannot find symbol
@Mapper(componentModel = "spring", uses = {UserMapper.class})
                                           ^
  symbol: class UserMapper
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/mapper/AMapper.java:11: error: Couldn't retrieve @Mapper annotation
@Mapper(componentModel = "spring", uses = {UserMapper.class})
^
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.17.Final
6 errors
```


 
<!-- Explain why this is a bug for you -->

##### **Reproduce the error**

<!-- For bug reports, an unambiguous set of steps to reproduce the error -->

##### **Related issues**

<!-- Has a similar issue been reported before? Please search both closed & open issues -->

##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

```
microservicea@0.0.0 /Users/jsm174/git/jhipster/react-test-9/microservicea
└── generator-jhipster@5.4.2

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "clientPackageManager": "npm",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": [
      "en",
      "fr"
    ],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.example.microservicea",
    "packageFolder": "com/example/microservicea",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "serviceDiscoveryType": "eureka",
    "skipClient": true,
    "testFrameworks": [],
    "websocket": false,
    "baseName": "microservicea",
    "applicationType": "microservice",
    "authenticationType": "oauth2",
    "buildTool": "gradle",
    "jhipsterVersion": "5.4.2",
    "skipUserManagement": true,
    "cacheProvider": "hazelcast",
    "serverPort": "8081"
  },
  "entities": [
    "A",
    "B"
  ]
}
</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity A {
  data String
}
entity B {
  data String
}
relationship ManyToOne {
  A{user(login)} to User
}

dto A, B with mapstruct
service A, B with serviceClass
microservice A, B with microservicea

</pre>
</details>


##### **Environment and Tools**

java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.17.1 (Apple Git-112)

node: v10.11.0

npm: 6.4.1

yarn: 1.10.1

Docker version 18.06.1-ce, build e68fc7a

docker-compose version 1.22.0, build f46880f

-   [X] Checking this box is mandatory (this is just to show you read everything)

<!-- Love JHipster? Please consider supporting our collective:
👉  https://opencollective.com/generator-jhipster/donate -->

--- a/generators/entity-server/templates/src/main/java/package/common/inject_template.ejs
+++ b/generators/entity-server/templates/src/main/java/package/common/inject_template.ejs
@@ -22,9 +22,6 @@
         if (queryService) {
             beans.push({class: `${entityClass}QueryService`, instance: `${entityInstance}QueryService`});
         }
-        if (saveUserSnapshot) {
-            beans.push({class: 'UserRepository', instance: 'userRepository'});
-        }
     } else {
         beans.push({class: `${entityClass}Repository`, instance: `${entityInstance}Repository`});
         if (dto === 'mapstruct') {
@@ -36,6 +33,9 @@
         if (reactiveRepositories) {
             beans.push({class: `${entityClass}ReactiveRepository`, instance: `${entityInstance}ReactiveRepository`});
         }
+    }
+    if (saveUserSnapshot && (viaService || constructorName.endsWith('Resource'))) {
+        beans.push({class: 'UserRepository', instance: 'userRepository'});
     }
 _%>
 

--- a/generators/entity-server/templates/src/main/java/package/common/inject_template.ejs
+++ b/generators/entity-server/templates/src/main/java/package/common/inject_template.ejs
@@ -16,53 +16,31 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 -%>
-<%_ if (!viaService) { _%>
+<%_ const beans = [];
+    if (viaService) {
+        beans.push({class: `${entityClass}Service`, instance: `${entityInstance}Service`});
+        if (queryService) {
+            beans.push({class: `${entityClass}QueryService`, instance: `${entityInstance}QueryService`});
+        }
+        if (saveUserSnapshot) {
+            beans.push({class: 'UserRepository', instance: 'userRepository'});
+        }
+    } else {
+        beans.push({class: `${entityClass}Repository`, instance: `${entityInstance}Repository`});
+        if (dto === 'mapstruct') {
+            beans.push({class: `${entityClass}Mapper`, instance: `${entityInstance}Mapper`});
+        }
+        if (searchEngine === 'elasticsearch') {
+            beans.push({class: `${entityClass}SearchRepository`, instance: `${entityInstance}SearchRepository`});
+        }
+        if (reactiveRepositories) {
+            beans.push({class: `${entityClass}ReactiveRepository`, instance: `${entityInstance}ReactiveRepository`});
+        }
+    }
+_%>
 
-    private final <%= entityClass %>Repository <%= entityInstance %>Repository;
-    <%_ if (dto === 'mapstruct') { _%>
+    <%= beans.map(bean => `private ${bean.class} ${bean.instance};`).join('\n\n    ') %>
 
-    private final <%= entityClass %>Mapper <%= entityInstance %>Mapper;
-    <%_ } _%>
-    <%_ if (searchEngine === 'elasticsearch') { _%>
-
-    private final <%= entityClass %>SearchRepository <%= entityInstance %>SearchRepository;
-    <%_ } _%>
-    <%_ if (reactiveRepositories) { _%>
-
-    private final <%= entityClass %>ReactiveRepository <%= entityInstance %>ReactiveRepository;
-    <%_ } _%>
-<%_  } else { _%>
-
-    private final <%= entityClass %>Service <%= entityInstance %>Service;
-    <%_ if (queryService) {  _%>
-
-    private final <%= entityClass %>QueryService <%= entityInstance %>QueryService;
-    <%_ } _%>
-<%_ } _%>
-<%_ if (saveUserSnapshot) { _%>
-
-    private final UserRepository userRepository;
-<%_ } _%>
-
-    public <%= constructorName %>(<% if (!viaService) { %><%= entityClass %>Repository <%= entityInstance %>Repository<% if (dto === 'mapstruct') { %>, <%= entityClass %>Mapper <%= entityInstance %>Mapper<% } %><% if (saveUserSnapshot) { %>, UserRepository userRepository<% } %><% if (searchEngine === 'elasticsearch') { %>, <%= entityClass %>SearchRepository <%= entityInstance %>SearchRepository<% } %><% if (reactiveRepositories) { %>, <%= entityClass %>ReactiveRepository <%= entityInstance %>ReactiveRepository<% } %><%  } else { %><%= entityClass %>Service <%= entityInstance %>Service<% } %><% if (queryService && viaService) { %>, <%= entityClass %>QueryService <%= entityInstance %>QueryService<% } %>) {
-        <%_ if (!viaService) { _%>
-        this.<%= entityInstance %>Repository = <%= entityInstance %>Repository;
-            <%_ if (dto === 'mapstruct') { _%>
-        this.<%= entityInstance %>Mapper = <%= entityInstance %>Mapper;
-            <%_ } _%>
-            <%_ if (searchEngine === 'elasticsearch') { _%>
-        this.<%= entityInstance %>SearchRepository = <%= entityInstance %>SearchRepository;
-            <%_ } _%>
-            <%_ if (reactiveRepositories) { _%>
-        this.<%= entityInstance %>ReactiveRepository = <%= entityInstance %>ReactiveRepository;
-            <%_ } _%>
-        <%_  } else { _%>
-        this.<%= entityInstance %>Service = <%= entityInstance %>Service;
-            <%_ if (queryService) {  _%>
-        this.<%= entityInstance %>QueryService = <%= entityInstance %>QueryService;
-            <%_ } _%>
-        <%_ }
-        if (saveUserSnapshot) { _%>
-        this.userRepository = userRepository;
-        <%_ } _%>
+    public <%= constructorName %>(<%= beans.map(bean => `${bean.class} ${bean.instance}`).join(', ') %>) {
+        <%= beans.map(bean => `this.${bean.instance} = ${bean.instance};`).join('\n        ') %>
     }

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIntTest.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIntTest.java.ejs
@@ -352,7 +352,7 @@ _%>
     public void setup() {
         MockitoAnnotations.initMocks(this);
         <%_ if (service !== 'no') { _%>
-        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Service<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Service<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
         <%_ } else { _%>
         final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Repository<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (reactiveRepositories) { %>, <%= entityInstance %>ReactiveRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
         <%_ } _%>

--- a/generators/entity/index.js
+++ b/generators/entity/index.js
@@ -900,7 +900,10 @@ module.exports = class extends BaseGenerator {
                         }
                     }
                     context.saveUserSnapshot =
-                        context.applicationType === 'microservice' && context.authenticationType === 'oauth2' && context.hasUserField;
+                        context.applicationType === 'microservice' &&
+                        context.authenticationType === 'oauth2' &&
+                        context.hasUserField &&
+                        context.dto === 'no';
 
                     if (_.isUndefined(relationship.otherEntityNamePlural)) {
                         relationship.otherEntityNamePlural = pluralize(relationship.otherEntityName);

